plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -536,7 +552,114 @@ public ConnectorInsertTableHandle beginInsert(ConnectorSession session, Connecto
     @Override
     public ColumnHandle getDeleteRowIdColumnHandle(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        return new IcebergColumnHandle(primitiveColumnIdentity(0, "$row_id"), BIGINT, ImmutableList.of(), BIGINT, Optional.empty());
+        return getUpdateRowIdColumnHandle(session, tableHandle, ImmutableList.of());
+    }
+
+    @Override
+    public ColumnHandle getUpdateRowIdColumnHandle(ConnectorSession session, ConnectorTableHandle tableHandle, List<ColumnHandle> updatedColumns)
+    {
+        // prepare Iceberg metadata fields
+        List<Types.NestedField> icebergFields = new ArrayList<>();
+        List<RowType.Field> trinoFields = new ArrayList<>();
+        List<Integer> path = new ArrayList<>();
+        icebergFields.add(DELETE_FILE_PATH);
+        trinoFields.add(RowType.field(DELETE_FILE_PATH.name(), VARCHAR));
+        path.add(DELETE_FILE_PATH.fieldId());
+        icebergFields.add(DELETE_FILE_POS);
+        trinoFields.add(RowType.field(DELETE_FILE_POS.name(), BIGINT));
+        path.add(DELETE_FILE_POS.fieldId());
+
+        // for update, include all the non-update columns
+        if (!updatedColumns.isEmpty()) {
+            IcebergTableHandle table = (IcebergTableHandle) tableHandle;
+            Set<Integer> updatedFields = updatedColumns.stream()
+                    .map(IcebergColumnHandle.class::cast)
+                    .map(IcebergColumnHandle::getId)
+                    .collect(Collectors.toSet());
+            for (Types.NestedField column : table.getTable().schema().columns()) {
+                if (!updatedFields.contains(column.fieldId())) {
+                    icebergFields.add(column);
+                    trinoFields.add(RowType.field(column.name(), toTrinoType(column.type(), typeManager)));
+                    path.add(column.fieldId());
+                }
+            }
+        }
+
+        Types.NestedField icebergRowIdField = Types.NestedField.required(TRINO_ROW_ID_COLUMN_ID, TRINO_ROW_ID_COLUMN_NAME, Types.StructType.of(icebergFields));
+        io.trino.spi.type.Type trinoRowIdType = RowType.from(trinoFields);
+        return new IcebergColumnHandle(createColumnIdentity(icebergRowIdField), trinoRowIdType, ImmutableList.of(), trinoRowIdType, Optional.empty());
+    }
+
+    @Override
+    public ConnectorTableHandle beginUpdate(ConnectorSession session, ConnectorTableHandle tableHandle, List<ColumnHandle> updatedColumns)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
```

Comment:
This should throw an unsupported exception if the table is v1 right? Same for `beginDelete`